File: code/ch05/5.1.1-path-sum.cpp
 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(root == NULL) return false;

        if(root->left == NULL && root->right == NULL) {
            if(root->val == sum) return true;
        
            return false;
        }

        return hasPathSum(root->left, sum - root->val) || hasPathSum(root->right, sum - root.val);
    }

};

File: code/ch05/5.1.1-path-sum.java
public class TreeNode {
  int val;
  TreeNode left;
  TreeNode right;

  TreeNode(int x) {
    val = x;
  }
}

class Solution {
  public boolean hasPathSum(TreeNode root, int sum) {
    if (root == null) return false;

    sum -= root.val;

    if ((root.left == null) && (root.right == null)) return (sum == 0);

    return hasPathSum(root.left, sum) || hasPathSum(root.right, sum);
  }
}